### PR TITLE
Add mangling for fixed point types proposed in N1169.

### PR DESCRIPTION
**Pull request by PiJoules**

### CODE REVIEW

**Comment by PiJoules** [comment]
Not sure how to "reference" issues with PRs, but this is for issue https://github.com/itanium-cxx-abi/cxx-abi/issues/56

File: abi.html [hunk 1]
```diff
@@ -612,6 +612,13 @@ <h3><a href="#docs"> 1.5 Base Documents </a></h3>
 References herein to the "C++ Standard," or to just the "Standard,"
 are to this document.
 
+<p>
+<li> [N1169]
+The <b>Programming languages - C - Extensions to support embedded processors</b>,
+ISO/IEC JTC1 SC22 WG14 N1169.
+This document describes extentions to the C language for supporting embedded
+processors, notabely including support for fixed point arithmetic.
```
Comment (zygoloid) [inline abi.html:620]:
```suggestion
processors, notably including support for fixed point arithmetic.
```

Comment (PiJoules) [reply abi.html]: Updated

File: abi.html [hunk 2]
```diff
@@ -5136,7 +5143,17 @@ <h5><a href="#mangling-builtin">5.1.5.2 Builtin types</a></h5>
                  ::= Da # auto
                  ::= Dc # decltype(auto)
                  ::= Dn # std::nullptr_t (i.e., decltype(nullptr))
+                 ::= [DS] DA <fixed-point-size> # [_Sat] T _Accum
```
Comment (rjmccall) [inline abi.html:5146]:
```suggestion
                 ::= [DS] DA <fixed-point-size> # N1169 fixed-point [_Sat] T _Accum
```

File: abi.html [hunk 3]
```diff
@@ -5136,7 +5143,17 @@ <h5><a href="#mangling-builtin">5.1.5.2 Builtin types</a></h5>
...
+                 ::= [DS] DR <fixed-point-size> # [_Sat] T _Fract
```
Comment (rjmccall) [inline abi.html:5147]:
```suggestion
                 ::= [DS] DR <fixed-point-size> # N1169 fixed-point [_Sat] T _Fract
```

**Review by rjmccall — CHANGES_REQUESTED** [review]
Generally looks good to me; just a couple minor editorial requests.

We have an active PR in #159 to make certain extended types substitutable.  Should these be substitution candidates if the same type is used in multiple places?

File: abi.html (hunk 3 above)
Comment (PiJoules) [reply abi.html]: Done

File: abi.html (hunk 2 above)
Comment (PiJoules) [reply abi.html]: Done

**Comment by PiJoules** [comment]
> We have an active PR in #159 to make certain extended types substitutable. Should these be substitution candidates if the same type is used in multiple places?

I asked around and we unfortunately don't know how common it would be to have mangled signatures with fixed-point types in them in places we might intend to use them. Sorry for the incomplete answer. Although given that the largest mangling for a fixed-point type is 5 characters and substitutions should be less than that, we still think having a sub should provide savings.

**Comment by rjmccall** [comment]
Okay.  The abstract guideline here is that adding substitutions does have a cost: in longer symbols, it can actually increase overall symbol size by filling the substitution table with a lot of extra types, making later substitutions larger.  I would say the line is around 4 bytes — a 3-byte mangling is probably not worth substituting, but a 5-byte mangling probably is.

Applying that to this would make only the saturating types candidates.  But treating different fixed-point types differently might be annoying to implement, and it's probably not a big deal either way.

Given that, do you have a strong opinion?

**Comment by PiJoules** [comment]
*ping* for any more comments on this

**Comment by zygoloid** [comment]
Looks good to me.

**Review by PiJoules — COMMENTED** [review]
Done

**Comment by rjmccall** [comment]
Alright, I think this has been open for long enough.

**Pull request merged**